File: oneflow/python/ops/two_stage_reduce.py
"""
Copyright 2020 The OneFlow Authors. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""
from __future__ import absolute_import

from typing import Optional, Sequence, Union

import oneflow as flow
import oneflow.python.framework.distribute as distribute_util
import oneflow.python.framework.hob as hob
import oneflow.python.framework.id_util as id_util
import oneflow.python.framework.remote_blob as remote_blob_util
import oneflow.python.lib.core.enable_if as enable_if
import oneflow.python.ops.user_op_builder as user_op_builder
from oneflow.python.oneflow_export import oneflow_export
import oneflow_api


@oneflow_export("math.two_stage_reduce_max")
def api_two_stage_reduce_max(
    x: oneflow_api.BlobDesc,
    axis: Optional[Union[int, Sequence[int]]] = None,
    keepdims: bool = False,
    name: Optional[str] = None,
) -> oneflow_api.BlobDesc:
    func = enable_if.unique([two_stage_reduce_max])
    return func(x, axis=axis, keepdims=keepdims, name=name)


@enable_if.condition(hob.in_global_mode)
def two_stage_reduce_max(x, axis=None, keepdims=False, name=None):
    name = name if name is not None else id_util.UniqueStr("ReduceMax_")
    return two_stage_reduce(x, axis, keepdims, "reduce_max", name)


@oneflow_export("math.two_stage_reduce_min")
def api_two_stage_reduce_min(
    x: oneflow_api.BlobDesc,
    axis: Optional[Union[int, Sequence[int]]] = None,
    keepdims: bool = False,
    name: Optional[str] = None,
) -> oneflow_api.BlobDesc:
    func = enable_if.unique([two_stage_reduce_min])
    return func(x, axis=axis, keepdims=keepdims, name=name)


@enable_if.condition(hob.in_global_mode)
def two_stage_reduce_min(x, axis=None, keepdims=False, name=None):
    name = name if name is not None else id_util.UniqueStr("ReduceMin_")
    return two_stage_reduce(x, axis, keepdims, "reduce_min", name)


def two_stage_reduce(x, axis=None, keepdims=False, op_type_name=None, name=None):

    assert check_x_dictribute(x, axis)
    axis = _check_axis(axis, x.shape)

    device_stage_out_list = []
    device_stage_count_list = []
    distribute_axis = x.distribute.axis
    x_list = flow.advanced.distribute_split(x, axis=distribute_axis)
    parallel_desc_symbol = flow.current_scope().device_parallel_desc_symbol
    device_tag = parallel_desc_symbol.device_tag
    parallel_id = 0
    for (
        machine_id,
        device_ids,
    ) in parallel_desc_symbol.machine_id2device_id_list.items():
        for device_id in device_ids:
            with flow.scope.placement(
                device_tag, str(machine_id) + ":" + str(device_id)
            ):
                device_stage_out, device_stage_count = reduce_device_stage(
                    x_list[parallel_id],
                    axis,
                    op_type_name + "_device_stage",
                    name + "_device_stage" + str(parallel_id),
                )
                device_stage_out_list.append(device_stage_out)
                device_stage_count_list.append(device_stage_count)
                parallel_id += 1
    device_stage_out = flow.advanced.distribute_concat(
        device_stage_out_list, axis=distribute_axis
    )
    device_stage_count = flow.advanced.distribute_concat(
        device_stage_count_list, axis=distribute_axis
    )

    device_stage_out = device_stage_out.with_distribute(flow.distribute.broadcast())
    device_stage_count = device_stage_count.with_distribute(flow.distribute.broadcast())

    out = reduce_global_stage(
        device_stage_out,
        device_stage_count,
        axis,
        keepdims,
        op_type_name + "_global_stage",
        name + "_global_stage",
    )
    return out


def reduce_device_stage(x, axis, op_name, name):
    out, mask, count = (
        flow.user_op_builder(name)
        .Op(op_name)
        .Input("in", [x])
        .Output("out")
        .Output("mask")
        .Output("count")
        .Attr("axis", axis)
        .Build()
        .InferAndTryRun()
        .RemoteBlobList()
    )
    return out, count


def reduce_global_stage(x, device_count, axis, keepdims, op_name, name):
    out, mask = (
        flow.user_op_builder(name)
        .Op(op_name)
        .Input("in", [x])
        .Input("device_count", [device_count])
        .Output("out")
        .Output("mask")
        .Attr("axis", axis)
        .Attr("keepdims", keepdims)
        .Build()
        .InferAndTryRun()
        .RemoteBlobList()
    )
    return out


def _check_axis(axis, shape):
    if axis is None:
        axis = list(range(len(shape)))

    if isinstance(axis, int):
        axis = [axis]

    assert isinstance(axis, (list, tuple)), "Invalid axis {}".format(axis)
    for x in axis:
        if x < 0:
            x += len(shape)
        assert x >= 0 and x < len(shape), "Invalid axis {}".format(axis)

    return axis


def check_x_dictribute(x, axis):
    for i in axis:
        if x.distribute is oneflow_api.sbp_descriptor.split(i):
            return True
    return False

File: oneflow/core/autograd/gradient_funcs/two_stage_reduce.cpp
/*
Copyright 2020 The OneFlow Authors. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#include "oneflow/core/framework/attr_map.h"
#include "oneflow/core/framework/op_expr_grad_function.h"
#include "oneflow/core/functional/functional.h"

namespace oneflow {
namespace one {

enum class ReduceMode : int32_t {
  kMin = 0,
  kMax = 1,
};

struct ReduceDeviceCaptureState : public AutoGradCaptureState {
  std::vector<int32_t> axis;
  bool requires_grad = false;
  size_t mask_index = -1;
  size_t count_index = -1;
};

template<ReduceMode mode>
class ReduceDevice : public OpExprGradFunction<ReduceDeviceCaptureState> {
 public:
  Maybe<void> Init(const OpExpr& op) override {
    const auto* op_expr = dynamic_cast<const UserOpExpr*>(&op);
    CHECK_NOTNULL_OR_RETURN(op_expr);  // NOLINT(maybe-need-error-msg)
    base_attrs_ = MakeAttrMapFromUserOpConf(op_expr->proto());
    return Maybe<void>::Ok();
  }

  Maybe<void> Capture(ReduceDeviceCaptureState* ctx, const TensorTuple& inputs,
                      const TensorTuple& outputs, const AttrMap& attrs) const override {
    CHECK_EQ_OR_RETURN(inputs.size(), 1);  // NOLINT(maybe-need-error-msg)
    ctx->requires_grad = inputs.at(0)->requires_grad();
    if (!ctx->requires_grad) { return Maybe<void>::Ok(); }

    ComposedAttrMap composed_attrs(attrs, base_attrs_);
    ctx->axis = JUST(composed_attrs.GetAttr<std::vector<int32_t>>("axis"));
    ctx->mask_index = ctx->SaveTensorForBackward(outputs.at(1));   // mask
    ctx->count_index = ctx->SaveTensorForBackward(outputs.at(2));  // count
    return Maybe<void>::Ok();
  }

  Maybe<void> Apply(const ReduceDeviceCaptureState* ctx, const TensorTuple& out_grads,
                    TensorTuple* in_grads) const override {
    if (!ctx->requires_grad) { return Maybe<void>::Ok(); }

    CHECK_EQ_OR_RETURN(out_grads.size(), 3);  // NOLINT(maybe-need-error-msg)
    const auto& mask = ctx->SavedTensors().at(ctx->mask_index);
    const auto& count = ctx->SavedTensors().at(ctx->count_index);
    in_grads->resize(1);
    if (mode == ReduceMode::kMin) {
      in_grads->at(0) =
          JUST(functional::ReduceMinDeviceStageGrad(out_grads.at(0), mask, count, ctx->axis));
    } else {
      in_grads->at(0) =
          JUST(functional::ReduceMaxDeviceStageGrad(out_grads.at(0), mask, count, ctx->axis));
    }
    return Maybe<void>::Ok();
  }

 private:
  AttrMap base_attrs_;
};

REGISTER_OP_EXPR_GRAD_FUNCTION("reduce_min_device_stage", ReduceDevice<ReduceMode::kMin>);
REGISTER_OP_EXPR_GRAD_FUNCTION("reduce_max_device_stage", ReduceDevice<ReduceMode::kMax>);

struct ReduceGlobalCaptureState : public AutoGradCaptureState {
  std::vector<int32_t> axis;
  bool requires_grad = false;
  bool keepdims = false;
  size_t mask_index = -1;
  size_t device_count_index = -1;
};

template<ReduceMode mode>
class ReduceGlobal : public OpExprGradFunction<ReduceGlobalCaptureState> {
 public:
  Maybe<void> Init(const OpExpr& op) override {
    const auto* fw_op_expr = dynamic_cast<const UserOpExpr*>(&op);
    CHECK_NOTNULL_OR_RETURN(fw_op_expr);  // NOLINT(maybe-need-error-msg)
    base_attrs_ = MakeAttrMapFromUserOpConf(fw_op_expr->proto());
    return Maybe<void>::Ok();
  }

  Maybe<void> Capture(ReduceGlobalCaptureState* ctx, const TensorTuple& inputs,
                      const TensorTuple& outputs, const AttrMap& attrs) const override {
    CHECK_EQ_OR_RETURN(inputs.size(), 2);   // NOLINT(maybe-need-error-msg)
    CHECK_EQ_OR_RETURN(outputs.size(), 2);  // NOLINT(maybe-need-error-msg)
    ctx->requires_grad = inputs.at(0)->requires_grad();
    if (!ctx->requires_grad) { return Maybe<void>::Ok(); }

    ComposedAttrMap composed_attrs(attrs, base_attrs_);
    ctx->axis = JUST(composed_attrs.GetAttr<std::vector<int32_t>>("axis"));
    ctx->keepdims = JUST(composed_attrs.GetAttr<bool>("keepdims"));
    ctx->mask_index = ctx->SaveTensorForBackward(outputs.at(1));         // mask
    ctx->device_count_index = ctx->SaveTensorForBackward(inputs.at(1));  // device_count
    return Maybe<void>::Ok();
  }

  Maybe<void> Apply(const ReduceGlobalCaptureState* ctx, const TensorTuple& out_grads,
                    TensorTuple* in_grads) const override {
    if (!ctx->requires_grad) { return Maybe<void>::Ok(); }
    CHECK_EQ_OR_RETURN(out_grads.size(), 2);  // NOLINT(maybe-need-error-msg)
    const auto& mask = ctx->SavedTensors().at(ctx->mask_index);
    const auto& device_count = ctx->SavedTensors().at(ctx->device_count_index);
    in_grads->resize(2);
    if (mode == ReduceMode::kMin) {
      in_grads->at(0) = JUST(functional::ReduceMinGlobalStageGrad(
          out_grads.at(0), mask, device_count, ctx->axis, ctx->keepdims));
    } else {
      in_grads->at(0) = JUST(functional::ReduceMaxGlobalStageGrad(
          out_grads.at(0), mask, device_count, ctx->axis, ctx->keepdims));
    }
    return Maybe<void>::Ok();
  }

 private:
  AttrMap base_attrs_;
};

REGISTER_OP_EXPR_GRAD_FUNCTION("reduce_min_global_stage", ReduceGlobal<ReduceMode::kMin>);
REGISTER_OP_EXPR_GRAD_FUNCTION("reduce_max_global_stage", ReduceGlobal<ReduceMode::kMax>);

}  // namespace one
}  // namespace oneflow
